Payment Formula Edited, Hints added for Formula, formatting Hints moved up above random number form (#1)
appdev-lessons/sinatra-omnicalc-1: content.md

</div>

## Objectives

Here is how the app should work. As a user:

- If I visit the path **/square/new**, I should see a form with a label and an input to enter a number.
  - If I submit that form, I should see the square of the number that I entered.
- If I visit the path **/square_root/new**, I should see a form with a label and an input to enter a number.
  - If I submit that form, I should see the square root of the number that I entered.
- If I visit the path **/payment/new**, I should see a form with labels and inputs to enter three values:

  - The APR (annual percentage rate).
  - The number of _years_ remaining.
  - The present value.
  - If I submit that form, I should see the **monthly** payment due given the values that I entered.
  - Mind your units! Use this formula:

      <!-- ![Payment formula](assets/omnical-1/payment_formula.gif) -->

    ![](assets/payments-formula.png)

<div class="bg-blue-100 py-1 px-5" markdown="1">

**Calculating Monthly Payments: Helpful Tips**

When calculating monthly payments, it's important to keep these hints in mind:

**1. Handling APR (Annual Percentage Rate):**

> - The 'r' in the formula represents the interest rate per period. We take the APR(Annual Percentage Rate) as a string input, so follow these steps:

Reviewer: ```suggestion
- The "r" in the formula represents the interest rate per period. We take the APR(Annual Percentage Rate) as a string input, so follow these steps:
```

Here and elsewhere use double quotes when referencing the formula parameters

Author: This is completed

> - The 'r' in the formula represents the interest rate per period. We take the APR(Annual Percentage Rate) as a string input, so follow these steps:

Reviewer: ```suggestion
- The 'r' in the formula represents the interest rate per period. We take the APR (Annual Percentage Rate) as a string input, so follow these steps:
```

>   a. Convert the user input from a string to a float.

Reviewer: Your a, b, c lists in the hints are not rendering on new lines in the rendered lesson. I suggest changing these to simple un-ordered lists without the lettering:

```
- Convert ...
- Divide ...
- etc.
```

Author: This is completed - I'm not sure if I should resolve the conversation or if it is better for the reviewer to resolve it.

Reviewer: I will resolve them.

  b. Divide the resulting number by 100 to convert from a percentage to a rate.
  c. Since we are calculating monthly payments, divide by 12 as well.

**2. Managing the Loan Term ('n' in the Formula):**

- The 'n' in the formula stands for the number of periods, which is typically provided by users in years. To adapt it for monthly payments, do the following:
  a. Convert the user input from a string to an integer.
  b. Multiply 'n' by 12 to calculate the number of monthly periods.

**3. Variable Organization:**

- To facilitate debugging and monitoring, create two variables: one for the numerator and another for the denominator.

> - If these are instance variables, you can access and view them within your program or view page.

Reviewer: ```suggestion
- If these are instance variables (i.e. `@numerator` with the leading `@`), then you can access and view them within your `.erb` view templates.
```

Author: This is completed

In particular these two:

- [`.to_fs(:currency)`](https://learn.firstdraft.com/lessons/33-the-one-ruby-reference#currency)
- [`.to_fs(:percentage, {:precision => 4})`](https://learn.firstdraft.com/lessons/33-the-one-ruby-reference#percentage)

> 

Reviewer: ```suggestion

<div class="bg-blue-100 py-1 px-5" markdown="1">

Use the `precision` option on the `to_fs` method to round your results. If you are confused by the usage, then please click the links above to read more about these methods.
</div>

```

Author: This is completed

could be useful when formatting the output of the payment form.

> 

Reviewer: ```suggestion

<div class="bg-blue-100 py-1 px-5" markdown="1">

Use the `precision` option on the `to_fs` method to round your results. If you are confused by the usage, then please click the links above to read more about these methods.
</div>

```

- If I visit the path **/random/new**, I should see a form with labels and inputs to enter two numbers, a minimum and a maximum.
  - If I submit that form, I should see a random number that falls between the numbers that I entered.

You can compare your app against [the target app](http://omnicalc-1.matchthetarget.com/), including doing "View Source" to look at some of the static HTML.

### Valid, Accessible Forms

**In order for your tests to pass**, you must build _valid_ forms (your Chrome browser _may_ tolerate invalid forms while you are manually testing, but automated test suites reject invalid forms):

- Each `<input>` in the form must have a unique `id=""` attribute.
- The `<label>` associated with the `<input>` should have a `for=""` attribute that matches the value of the `<input>`'s `id`.
- The copy within the `<label>` must exactly match the target — spelling, capitalization, and punctuation matter for labels.
- The same goes for the copy on the button to submit the form.
- Any invalid HTML within a form will cause the test to fail, e.g. an orphaned closing `</div>` tag. Keep your code neatly indented to help avoid this.

> An example of a valid form; in particular, notice the `id=""` and `for=""` attributes:

Reviewer: Cool feature of our special markdown in learn:

You can write

```
```erb{Line:(ColumnX-ColumnY)}
some code
```
```

To highlight specific lines and columns. Give it a try and view the results of your change in the rendered lesson (https://learn.firstdraft.com/lessons/232-sinatra-omnicalc-1). 

For instance, in the code block below I might write:

```
```erb{3:(12-26),7:(12-25)}
```

See how that highlights the `for=""` and `id=""`?

Author: Oh yeah - I just tried this - it's pretty cool!

Author: I also did the same for other id and for tags as well

Author: This is completed

---

- Approximately how long (in minutes) did this lesson take you to complete?

>   {: .free_text_number #time_taken title="Time taken" points="1" answer="any" }

Author: Yeah - this one is interesting since it autoformats on save even if I delete the space - should I disable and save?

>   {: .free_text_number #time_taken title="Time taken" points="1" answer="any" }

Author: Turned of auto-format and saved it. This is completed
